src/Crypto/Macaroon/Verifier/Internal.hs

```diff
@@ -72,26 +58,34 @@ verifySig k m = bool (Left SigMismatch) (Right m) $
     derivedKey = toBytes (hmac "macaroons-key-generator" k :: HMAC SHA256)
 
 -- | Given a list of verifiers, verify each caveat of the given macaroon
-verifyCavs :: (Functor m, MonadIO m)
+verifyCavs :: forall m. (Applicative m)
            => [Caveat -> m VerifierResult]
            -> Macaroon
            -> m (Either ValidationError Macaroon)
-verifyCavs verifiers m = gatherEithers <$> mapM validateCaveat (caveats m)
+verifyCavs verifiers m = toEither <$> errors
   where
-    {-
-     - validateCaveat :: Caveat -> m (Validation String Caveat)
-     - We can use fromJust here safely since we use a `Just Failure` as a
-     - starting value for the foldM. We are guaranteed to have a `Just something`
-     - from it.
-     -}
-    validateCaveat c = fmap (const c) . fromJust <$> foldM (\res v -> mappend res . fmap eitherToValidation . vResult2MaybeEither <$> v c) (defErr c) verifiers
-    -- vResult2MaybeEither :: VerifierResult -> Maybe (Either ValidationError ())
-    vResult2MaybeEither Unrelated = Nothing
-    vResult2MaybeEither Verified = Just (Right ())
-    vResult2MaybeEither (Refused e)= Just (Left e)
-    -- defErr :: Caveat -> Maybe (Validation String Caveat)
-    defErr c = Just $ Failure NoVerifier
-    -- gatherEithers :: [Validation String Caveat] -> Either String Caveat
-    gatherEithers vs = case partitionEithers . map validationToEither $ vs of
-        ([],_) ->  Right m
-        (errs,_) -> Left (mconcat errs)
+    toEither = maybe (Right m) (Left . OutstandingCaveats) . nonEmpty
+    errors = fmap catMaybes $ traverse keepErrors $ caveats m
+
+    -- apply the verifiers to a caveat and only keep errors
+    keepErrors :: Caveat -> m (Maybe OutstandingCaveat)
+    keepErrors =
+        fmap sequenceA . -- Move the Maybe outside the tuple
+        sequenceA . -- Move the m outside the tuple
+        getTaggedErrors
+
+    -- annotated results
+    getTaggedErrors :: Caveat -> (Caveat, m (Maybe [VerifierError]))
+    getTaggedErrors = id &&& fmap collapseErrors . applyVerifiers
+
+    -- collapse all the results, @Nothing@ means the caveat has been
+    -- verified
+    collapseErrors :: [VerifierResult] -> Maybe [VerifierError]
+    collapseErrors rs =
+      if Verified `elem` rs
+        then Nothing
```

Comment:
If I follow this correctly, this means that `Verified` is an annihilator, e.g. one successful verifier is stronger than all the errors in the others.

This is an important behaviour IMHO, and should be somewhere in the docs

Reply:
Not doing so would be a departure from all the other macaroons implentations, where `Unrelated` and `VerifierError` are collapsed under `false`.

Reply:
I'm not saying that this should be changed. I was confused by all the other implementations as well, especially that everything is crammed into a boolean.
It's more a docs issue, as I'm looking forward to shipping a 1.0 release.